codespan-reporting/src/diagnostic.rs

```diff
@@ -113,45 +121,148 @@ impl Diagnostic {
+    pub fn new_note(
+        message: impl Into<String>,
+        primary_label: Label<FileId>,
+    ) -> Diagnostic<FileId> {
         Diagnostic::new(Severity::Note, message, primary_label)
     }
 
     /// Create a new diagnostic with a severity of `Severity::Help`.
-    pub fn new_help(message: impl Into<String>, primary_label: Label) -> Diagnostic {
+    pub fn new_help(
+        message: impl Into<String>,
+        primary_label: Label<FileId>,
+    ) -> Diagnostic<FileId> {
         Diagnostic::new(Severity::Help, message, primary_label)
     }
 
     /// Add an error code to the diagnostic.
-    pub fn with_code(mut self, code: impl Into<String>) -> Diagnostic {
+    pub fn with_code(mut self, code: impl Into<String>) -> Diagnostic<FileId> {
         self.code = Some(code.into());
         self
     }
 
     /// Add some notes to the diagnostic.
-    pub fn with_notes(mut self, notes: Vec<String>) -> Diagnostic {
+    pub fn with_notes(mut self, notes: Vec<String>) -> Diagnostic<FileId> {
         self.notes = notes;
         self
     }
 
     /// Add some secondary labels to the diagnostic.
-    pub fn with_secondary_labels(mut self, labels: impl IntoIterator<Item = Label>) -> Diagnostic {
+    pub fn with_secondary_labels(
+        mut self,
+        labels: impl IntoIterator<Item = Label<FileId>>,
+    ) -> Diagnostic<FileId> {
         self.secondary_labels.extend(labels);
         self
     }
 }
+
+/// A location in a source file.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Location {
+    /// The line index in the source file.
+    pub line: usize,
+    /// The column index in the source file.
+    pub column: usize,
+}
+
+/// A line within a source file.
+pub struct Line<Source> {
+    /// The starting byte index of the line.
+    pub start: usize,
+    /// The source of the line.
+    pub source: Source,
+}
+
+impl<Source> Line<Source>
+where
+    Source: AsRef<str>,
+{
+    pub fn column_index(&self, byte_index: usize) -> Option<usize> {
+        match byte_index.checked_sub(self.start) {
+            None => Some(0),
+            Some(relative_index) => {
+                let source = self.source.as_ref().get(..relative_index)?;
+                Some(source.chars().count())
+            },
+        }
+    }
+}
+
+/// Files that can be used for pretty printing.
+pub trait Files {
+    type FileId: Copy + PartialEq + PartialOrd + Eq + Ord + std::hash::Hash;
+    type Origin: std::fmt::Display;
+    type LineSource: AsRef<str>;
+
+    /// The origin of a file.
+    fn origin(&self, id: Self::FileId) -> Option<Self::Origin>;
+
+    /// The line at the given index.
+    fn line(&self, id: Self::FileId, line_index: usize) -> Option<Line<Self::LineSource>>;
+
+    /// The index of the line at the given line index.
+    fn line_index(&self, id: Self::FileId, byte_index: usize) -> Option<usize>;
+
+    /// The location of the given byte index.
```

Comment:
```suggestion
    /// A user-facing location representing the given byte index.
```

Reply:
Whoops, misread the API - this is not user facing, it just has the `(line_index, column_index)` pair.